src/StreamJsonRpc.Tests/JsonRpcWithFatalExceptionsTests.cs

```diff
@@ -0,0 +1,323 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio.Threading;
+using Nerdbank;
+using StreamJsonRpc;
+using Xunit;
+using Xunit.Abstractions;
+
+public class JsonRpcWithFatalExceptionsTests : TestBase
+{
+    internal static Exception FaultException;
+    private readonly Server server;
+    private readonly DisposingDuplexStream serverStream;
+    private readonly DisposingDuplexStream clientStream;
+    private readonly DelimitedMessageHandler messageHandler;
+    private readonly JsonRpc clientRpc;
+
+    public JsonRpcWithFatalExceptionsTests(ITestOutputHelper logger)
+        : base(logger)
+    {
+        this.server = new Server();
+        var streams = DisposingDuplexStream.CreateStreams();
+        this.serverStream = streams.Item1;
+        this.clientStream = streams.Item2;
+
+        this.messageHandler = new DisposingMessageHander(this.clientStream, this.serverStream);
+        this.clientRpc = new JsonRpcWithFatalExceptions(this.messageHandler, this.server);
+        this.clientRpc.StartListening();
+    }
+
+    [Fact]
+    public async Task CanInvokeMethodOnServer()
+    {
+        string testLine = "TestLine1" + new string('a', 1024 * 1024);
+        string result1 = await this.clientRpc.InvokeAsync<string>(nameof(Server.ServerMethod), testLine);
+        Assert.Equal(testLine + "!", result1);
+    }
+
+    [Fact]
+    public async Task StreamsStayOpenIfCancelledTaskReturned()
+    {
+        var ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => this.clientRpc.InvokeAsync(nameof(Server.ServerMethodThatReturnsCancelledTask)));
+        Assert.Equal(CancellationToken.None, ex.CancellationToken);
+
+        // Assert MessageHandler and Stream objects are not disposed
+        Assert.False(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
+        Assert.False(this.serverStream.IsDisposed);
+        Assert.False(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    public async Task CloseStreamsOnSynchronousMethodException()
+    {
+        var exceptionMessage = "Exception from CloseStreamsOnSynchronousMethodException";
+        OperationCanceledException exception = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => this.clientRpc.InvokeAsync(nameof(Server.MethodThatThrowsUnauthorizedAccessException), exceptionMessage));
+        Assert.NotNull(exception.StackTrace);
+        Assert.Equal(FaultException.InnerException.Message, exceptionMessage);
```

Comment:
The exception that's passed into the filter is `TargetInvocationException` in the synchronous case and `AggregateException` in the asynchronous case. Would you rather I document that, or use the inner exception in the filter?
(I'm leaning towards documenting, myself)

Reply:
@AArnott

Reply:
I'm leaning toward sharing the original exception rather than the exception we add as an artifact of how we invoke the method. 
For `AggregateException`, is that thrown because `Task.Exception` is thrown? What is the `InnerException`? Is it a `TargetInvocationException`?

I'd really like to get to a consistent exception type because I suspect folks who override this method will not anticipate that async vs. sync methods will need to be tested separately. If they're all `TargetInvocationException` or `AggregateException` that might be OK, but the mix of the two sounds like it's a recipe for overrides to be buggy. I'm guessing their first guess (at least in lieu of docs) would be that if they throw `ArgumentException` from their server method, that they'll see an `ArgumentException` in the exception filter. And IMO that seems like a reasonable goal for us, if we can reach it.